### PR TITLE
Alerts backend with metrics

### PR DESCRIPTION
**Pull request by mindhash**
This PR:
- introduces a new rule engine (pkg/query-service/rules) with a generic rule framework (supports threshold rule now but in future rules like anomaly detection can be added) and multiple rule backends (e.g. clickhouse queries, promql expressions)
- removes rule management dependencies from prometheus and clickhouse reader completely

Ref Issue: [#347](https://github.com/SigNoz/engineering-pod/issues/347)
frontend PR: [#1359](https://github.com/SigNoz/signoz/pull/1359)

### CODE REVIEW

**Review by srikanthccv — COMMENTED** [review]
I tried three times to take look at this in the past several days but I can't bring myself to review 5k lines of addition and the logic. @mindhash help me understand better sometime today again.

File: pkg/query-service/app/http_handler.go [hunk 1]
```diff
@@ -615,13 +624,17 @@ func (aH *APIHandler) queryRangeMetricsV2(w http.ResponseWriter, r *http.Request
...
 
-func (aH *APIHandler) listRulesFromProm(w http.ResponseWriter, r *http.Request) {
-	alertList, apiErrorObj := (*aH.reader).ListRulesFromProm()
-	if apiErrorObj != nil {
-		respondError(w, apiErrorObj, nil)
+func (aH *APIHandler) listRules(w http.ResponseWriter, r *http.Request) {
+
+	rules, err := aH.ruleManager.ListRuleStates()
+	if err != nil {
+		respondError(w, &model.ApiError{Typ: model.ErrorBadData, Err: err}, nil)
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go:631]: I think generally this type is not being used in front-end. its just info only and useful to view from network tab. 
The frontend just throws 'something went wrong in all cases. thats how i found front-end being implemented. 
I am not sure if this type is important. may be i am missing something?

File: pkg/query-service/app/http_handler.go [hunk 2]
```diff
@@ -757,32 +770,36 @@ func (aH *APIHandler) createDashboards(w http.ResponseWriter, r *http.Request) {
...
 	id := mux.Vars(r)["id"]
+	fmt.Println("deleting rule:", id)
 
-	apiErrorObj := (*aH.reader).DeleteRule(id)
+	err := aH.ruleManager.DeleteRule(id)
 
-	if apiErrorObj != nil {
-		respondError(w, apiErrorObj, nil)
+	if err != nil {
+		respondError(w, &model.ApiError{Typ: model.ErrorInternal, Err: err}, nil)
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go:780]: the error handling is similar to how earlier code was handling it. I am not sure how important error type is. I am assuming for dev errors, the logs will be key to getting the right context. 
The UI does very minimal in terms of api errors. unless there are validation errors offcourse

File: pkg/query-service/app/server.go [hunk 3]
```diff
@@ -315,3 +383,49 @@ func (s *Server) Start() error {
...
+
+	// create manager opts
+	managerOpts := &rules.ManagerOptions{
+		NotifierOpts: notifierOpts,
+		Queriers: &rules.Queriers{
+			PqlEngine: pqle,
+			Ch:        ch.GetConn(),
+		},
+		ExternalURL:  externalURL,
+		Conn:         db,
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/server.go:372]: named it DBConn .. as the db type can always change (e.g. postgres) .. its a sqlx.DB conn

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 4]
```diff
@@ -548,156 +336,8 @@ func connect(cfg *namespaceConfig) (clickhouse.Conn, error) {
...
-	groupName := fmt.Sprintf("%d-groupname", rule.Id)
-
-	err := r.ruleManager.AddGroup(time.Duration(r.promConfig.GlobalConfig.EvaluationInterval), rule.Data, groupName)
-
-	if err != nil {
-		return &model.ApiError{Typ: model.ErrorInternal, Err: err}
-	}
-
-	return nil
+func (r *ClickHouseReader) GetConn() clickhouse.Conn {
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/clickhouseReader/reader.go:339]: I am not sure how else to handle it. I get ch connection only through clickhouseReader.

File: pkg/query-service/app/http_handler.go [hunk 5]
```diff
@@ -291,13 +298,14 @@ func (aH *APIHandler) RegisterPrivateRoutes(router *mux.Router) {
 func (aH *APIHandler) RegisterRoutes(router *mux.Router) {
 	router.HandleFunc("/api/v1/query_range", ViewAccess(aH.queryRangeMetrics)).Methods(http.MethodGet)
 	router.HandleFunc("/api/v1/query", ViewAccess(aH.queryMetrics)).Methods(http.MethodGet)
-	router.HandleFunc("/api/v1/channels", ViewAccess(aH.listChannels)).Methods(http.MethodGet)
+	// router.HandleFunc("/api/v1/channels", ViewAccess(aH.listChannels)).Methods(http.MethodGet)
+	router.HandleFunc("/api/v1/channels", aH.listChannels).Methods(http.MethodGet)
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go:302]: fixed. Thanks for pointing this out.

File: pkg/query-service/utils/labels/labels.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+)
+
+const sep = '\xff'
+
+// Well-known label names used by Prometheus components.
+const (
+	MetricName   = "__name__"
+	AlertName    = "alertname"
+	BucketLabel  = "le"
+	InstanceName = "instance"
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/utils/labels/labels.go:20]: changed to MetricNameLabel instead as thats what it is.

File: pkg/query-service/integrations/alertManager/notifier.go [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+	for {
+		select {
+		case <-n.ctx.Done():
+			return
+		case <-n.more:
+		}
+		alerts := n.nextBatch()
+
+		if !n.sendAll(alerts...) {
+			zap.S().Warn("msg: dropped alerts", "/t count:", len(alerts))
```
Comment (srikanthccv) [inline pkg/query-service/integrations/alertManager/notifier.go:136]: did you intend to use `\t`?

**Review by srikanthccv — COMMENTED** [review]
Couple of comments. Overall LGTM.

File: pkg/query-service/app/http_handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go:779]: updated to ErrorInternal

File: pkg/query-service/app/http_handler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go]: updated to ErrorInternal.  As the likely error here is due to rule state not being fetched or some code issue

**Review by srikanthccv — APPROVED** [review]

File: pkg/query-service/app/http_handler.go [hunk 8]
```diff
@@ -379,12 +387,13 @@ func Intersection(a, b []int) (c []int) {
 
 func (aH *APIHandler) getRule(w http.ResponseWriter, r *http.Request) {
 	id := mux.Vars(r)["id"]
-	alertList, apiErrorObj := (*aH.reader).GetRule(id)
-	if apiErrorObj != nil {
-		respondError(w, apiErrorObj, nil)
+	ruleResponse, err := aH.ruleManager.GetRule(id)
+	fmt.Println("ruleResponse", ruleResponse)
+	if err != nil {
+		respondError(w, &model.ApiError{Typ: model.ErrorBadData, Err: err}, nil)
```
[the remark this answers is not in the document]
Comment (mindhash) [reply pkg/query-service/app/http_handler.go]: changed to ErrorInternal type

**Pull request merged**